pallets/stateful-storage/src/lib.rs

```diff
@@ -260,17 +263,94 @@ pub mod pallet {
 
 		#[pallet::call_index(1)]
 		#[pallet::weight(0)]
-		pub fn upsert_page(_origin: OriginFor<T>, payload: Vec<u8>) -> DispatchResult {
+		pub fn upsert_page(
+			origin: OriginFor<T>,
+			state_owner_msa_id: MessageSourceId,
+			schema_id: SchemaId,
+			page_id: PageId,
+			payload: Vec<u8>,
+		) -> DispatchResult {
 			ensure!(
 				payload.len() as u32 <= T::MaxPaginatedPageSizeBytes::get(),
 				Error::<T>::PageExceedsMaxPageSizeBytes
 			);
+			ensure!(
+				page_id as u32 <= T::MaxPaginatedPageId::get(),
+				Error::<T>::PageIdExceedsMaxAllowed
+			);
+			let provider_key = ensure_signed(origin)?;
+			let provider_msa_id = T::MsaInfoProvider::ensure_valid_msa_key(&provider_key)
+				.map_err(|_| Error::<T>::InvalidMessageSourceAccount)?;
+
+			let schema = T::SchemaProvider::get_schema_by_id(schema_id);
+			ensure!(schema.is_some(), Error::<T>::InvalidSchemaId);
+			ensure!(
+				schema.unwrap().payload_location == PayloadLocation::Paginated,
+				Error::<T>::SchemaPayloadLocationMismatch
+			);
+
+			let current_block = frame_system::Pallet::<T>::block_number();
+			T::SchemaGrantValidator::ensure_valid_schema_grant(
+				ProviderId(provider_msa_id),
+				DelegatorId(state_owner_msa_id),
+				schema_id,
+				current_block,
+			)
+			.map_err(|_| Error::<T>::UnAuthorizedDelegate)?;
+
+			let schema_key: StatefulPageKeyPart = schema_id.encode();
+			let page_key: StatefulPageKeyPart = page_id.encode();
+
+			StatefulChildTree::write(&state_owner_msa_id, &[schema_key, page_key], payload);
+			Self::deposit_event(Event::PaginatedPageUpdated {
+				msa_id: state_owner_msa_id,
+				schema_id,
+				page_id,
+			});
 			Ok(())
 		}
 
 		#[pallet::call_index(2)]
 		#[pallet::weight(0)]
-		pub fn remove_page(_origin: OriginFor<T>) -> DispatchResult {
+		pub fn remove_page(
+			origin: OriginFor<T>,
+			state_owner_msa_id: MessageSourceId,
+			schema_id: SchemaId,
+			page_id: PageId,
+		) -> DispatchResult {
+			ensure!(
+				page_id as u32 <= T::MaxPaginatedPageId::get(),
+				Error::<T>::PageIdExceedsMaxAllowed
+			);
+			let provider_key = ensure_signed(origin)?;
+			let provider_msa_id = T::MsaInfoProvider::ensure_valid_msa_key(&provider_key)
+				.map_err(|_| Error::<T>::InvalidMessageSourceAccount)?;
+
+			let schema = T::SchemaProvider::get_schema_by_id(schema_id);
```

Comment:
Seems like this block of code can be refactored into a helper method.  Both extrinsic seem to be doing these checks. By moving into a helper method it allows for it to be unit tested.


```
fn ensure_valid_schema_something() {
		let schema = T::SchemaProvider::get_schema_by_id(schema_id);
			ensure!(schema.is_some(), Error::<T>::InvalidSchemaId);
			ensure!(
				schema.unwrap().payload_location == PayloadLocation::Paginated,
				Error::<T>::SchemaPayloadLocationMismatch
			);

			let current_block = frame_system::Pallet::<T>::block_number();
			T::SchemaGrantValidator::ensure_valid_schema_grant(
				ProviderId(provider_msa_id),
				DelegatorId(state_owner_msa_id),
				schema_id,
				current_block,
			)
			.map_err(|_| Error::<T>::UnAuthorizedDelegate)?;
			....
			}
```

Reply:
Nice idea; I like it. Actually I think the same treatment can be applied to most of the checks in the extrinsics for this pallet.